src/BloomExe/web/EnhancedImageServer.cs

```diff
@@ -1,51 +1,137 @@
-
+using Bloom.ImageProcessing;
 
 namespace Bloom.web
 {
 	/// <summary>
 	/// A local http server that can serve (low-res) images plus other files.
 	/// </summary>
+	/// <remarks>geckofx makes concurrent requests of URLs which this class handles. This means
+	/// that the methods of this class get called on different threads, so it has to be
+	/// thread-safe.</remarks>
 	public class EnhancedImageServer: ImageServer
 	{
 		private FileSystemWatcher _sampleTextsWatcher;
 		private bool _sampleTextsChanged = true;
+		private const int RWLockTimeOut = 100;
 
 		public CollectionSettings CurrentCollectionSettings { get; set; }
 
 		public EnhancedImageServer(LowResImageCache cache): base(cache)
 		{
+			Lock = new ReaderWriterLock();
+		}
+
+		// We use three different locks to synchronize access to the methods of this class.
+		// This allows certain methods to run concurrently.
+
+		// used to synchronize access to properties. Use this lock when you want to allow
+		// multiple concurrent readers, but no one else when writing the property.
+		private ReaderWriterLock Lock { get; set; }
+		// used to synchronize access to I18N methods
+		private object I18NLock = new object();
+		// used to synchronize access to various methods
+		private object SyncObj = new object();
+
+		private string _currentPageContent;
+		private string _accordionContent;
+		private bool _authorMode;
+
+		public string CurrentPageContent
+		{
+			get { return ReaderLock(() => _currentPageContent); }
+			set { WriterLock(() => _currentPageContent = value); }
 		}
 
-		public string CurrentPageContent { get; set; }
-		public string AccordionContent { get; set; }
-		public bool AuthorMode { get; set; }
+		public string AccordionContent
+		{
+			get { return ReaderLock(() => _accordionContent); }
+			set { WriterLock(() => _accordionContent = value); }
+		}
+
+		public bool AuthorMode
+		{
+			get { return ReaderLock(() => _authorMode); }
+			set { WriterLock(() => _authorMode = value); }
+		}
 
 		/// <summary>
-		/// There can really only be one of these globally, since ReadersHandler is static. But for now that's true anyway
-		/// because we use a fixed port. See comments on the ReadersHandler property.
+		/// There can really only be one of these globally, since ReadersHandler is static. But for
+		/// now that's true anyway because we use a fixed port. See comments on the ReadersHandler
+		/// property.
 		/// </summary>
 		public Book.Book CurrentBook
 		{
-			get { return ReadersHandler.CurrentBook; }
-			set { ReadersHandler.CurrentBook = value; }
+			get
+			{
+				lock (SyncObj)
+				{
+					return ReadersHandler.CurrentBook;
+				}
+			}
+			set
+			{
+				lock (SyncObj)
+				{
+					ReadersHandler.CurrentBook = value;
+				}
+			}
+		}
+
+		#region Reader/Writer Lock helper methods
+
+		private T ReaderLock<T>(Func<T> func)
+		{
+			try
+			{
+				Lock.AcquireReaderLock(RWLockTimeOut);
+				return func();
+			}
+			catch (ApplicationException e)
```

Comment:
Is there not a more specific exception we can catch for timeout? I don't suppose it matters if we only use this for reading and writing simple properties...no way that will either time out or throw...but I'd hate to convert some unrelated exception into a false timeout report.